hazelcast/src/main/java/com/hazelcast/instance/impl/Node.java

```diff
@@ -936,4 +1039,94 @@ private Map<String, String> findMemberAttributes(MemberAttributeConfig attribute
         }
         return attributes;
     }
+
+    @SuppressWarnings({"checkstyle:npathcomplexity", "checkstyle:cyclomaticcomplexity", "checkstyle:methodlength"})
+    @Override
+    public void update(int previousClusterSpecSize, int currentClusterSpecSize,
+                       int readyNodesCount, int currentNodesCount) {
+        if (previousClusterSpecSize == UNKNOWN) {
+            if (currentClusterSpecSize > 0
+                    && (readyNodesCount == UNKNOWN || readyNodesCount == 0)) {
+                // startup of first member of new cluster
+                logger.info("Cluster starting in managed context");
+                clusterTopologyState.set(ClusterTopologyState.of(CLUSTER_START, currentClusterSpecSize));
+            } else {
+                logger.info("Member starting in managed context");
+                clusterTopologyState.set(ClusterTopologyState.UNKNOWN);
+            }
+            return;
+        }
+        final ClusterTopologyState previousState = clusterTopologyState.get();
+        final ClusterTopologyIntent previous = previousState.getClusterTopologyIntent();
```

Comment:
To me it is hard to distinguish between state and intent when reading code? Can we do something for this? Maybe use one of them and delete other one or rename?

Reply:
agreed, refactored in 96f2f61ac69d4ac11bcca2b032de9fdfb03269cf. See also the added javadoc in 4f840b6f9e7a8b6f775e7ee9951358689ce5d022